File: Data Structures & Algorithms/reorder-linked-list/submission-1.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func reorderList(head *ListNode) {
	if head == nil {
		return
	}

	var prev *ListNode

	p := head
	p2 := head
    for {
		prev = p2
		p2 = p2.Next
		p = p.Next
		if (p == nil) {
			break
		}
		p = p.Next
		if (p == nil) {
			break
		}
	}

	prev.Next = nil

    p2 = reverseList(p2)

	p = head

	var t, t2 *ListNode
	for ;; {
		if p2 == nil {
			break
		}

		t = p.Next
		t2 = p2.Next
		p.Next = p2
		p2.Next = t
		p = t
		p2 = t2
	}
}

func reverseList(head *ListNode) *ListNode {
    h := head;

    if h == nil {
        return h
    }

    m := h.Next

    if m == nil {
        return h
    }

    h = nil
    m = head

    t := m.Next

    for {
        m.Next = h

        h = m
        m = t

        if t == nil {
            break
        }

        t = t.Next
    }

    return h
}

File: Data Structures & Algorithms/add-two-numbers/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode res = null;
        ListNode prev = null;
        ListNode p1 = l1;
        ListNode p2 = l2;
        boolean carry = false;
        for (;;) {
            int r = (p1 == null ? 0 : p1.val)
                + (p2 == null ? 0 : p2.val);
            if (carry) {
                r++;
            }

            ListNode n = new ListNode(r % 10);

            carry = r >= 10;

            if (res == null) {
                res = n;
            }

            if (prev != null) {
                prev.next = n;
            }

            prev = n;

            if (p1 != null) {
                p1 = p1.next;
            }

            if (p2 != null) {
                p2 = p2.next;
            }

            if (p1 == null && p2 == null) {
                if (carry) {
                    n = new ListNode(1);

                    prev.next = n;
                }

                break;
            }
        }

        return res;
    }
}
